### PR TITLE
feat(acp): add default agent profiles for claude and codex

### PR DESCRIPTION
**Pull request by siddseethepalli**
## Summary
- Add `DEFAULT_ACP_AGENT_PROFILES` for `claude` (claude-agent-acp) and `codex` (codex-acp) so enabling ACP yields useful defaults.
- Add `DEFAULT_AGENT_INSTALL_HINTS` keyed by command for reuse by the resolver and `acp_list_agents`.
- No consumers yet — wired in later PRs.

Part of plan: acp-codex-claude.md (PR 1 of 9)

---

<a href="https://app.devin.ai/review/vellum-ai/vellum-assistant/pull/28070" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by siddseethepalli** [comment]
<details>
<summary>Plan papertrail: ACP-based coding-agent management for Velissa (Codex + Claude Code)</summary>

# ACP-based coding-agent management for Velissa (Codex + Claude Code)

## Overview

Wire Velissa to delegate coding work to Codex and Claude Code via the existing ACP (Agent Client Protocol) infrastructure. Add `acp_steer` (interruptive) and `acp_list_agents` tools, default agent profiles for `claude` and `codex` so `acp.enabled: true` Just Works, a shared resolver with binary preflight, fixes for the deprecated Claude package name, and a non-blocking version-check. Use upstream `@zed-industries/codex-acp` (no vendored adapter). Source plan: `~/.claude/plans/i-want-velissa-to-recursive-eclipse.md`.

## PR 1: Add ACP default agent profiles

### Depends on
None

### Branch
acp-codex-claude/pr-1-default-profiles

### Title
feat(acp): add default agent profiles for claude and codex

### Files
- `assistant/src/config/acp-defaults.ts` (new)
- `assistant/src/config/acp-defaults.test.ts` (new)

### Implementation steps
1. Create `assistant/src/config/acp-defaults.ts`. Export `DEFAULT_ACP_AGENT_PROFILES: Readonly<Record<string, AcpAgentConfig>>` with two entries:
   - `claude`: `{ command: "claude-agent-acp", args: [], description: "Claude Code (via @agentclientprotocol/claude-agent-acp)" }`
   - `codex`: `{ command: "codex-acp", args: [], description: "OpenAI Codex CLI (via @zed-industries/codex-acp)" }`
   Import `AcpAgentConfig` from `./acp-schema.js`.
2. Also export `DEFAULT_AGENT_INSTALL_HINTS: Readonly<Record<string, string>>` keyed by command name (not agent id), so it can be reused later by the resolver and `acp_list_agents`:
   - `"claude-agent-acp"` → `"npm i -g @agentclientprotocol/claude-agent-acp"`
   - `"codex-acp"` → `"npm i -g @zed-industries/codex-acp"`
3. Add `assistant/src/config/acp-defaults.test.ts` that asserts the keys, command names, and freezes (use `Object.freeze` semantics — verify `as const` in TS gives readonly).

### Acceptance criteria
- `bunx tsc --noEmit` passes from `assistant/`.
- `bun test src/config/acp-defaults.test.ts` passes.
- No other file imports the new module yet (consumed in PR 4 and downstream).

## PR 2: Make ACP adapter version-check non-blocking

### Depends on
None

### Branch
acp-codex-claude/pr-2-version-check

### Title
refactor(acp): make adapter version-check non-blocking with timeout

### Files
- `assistant/src/tools/acp/spawn.ts`
- `assistant/src/tools/acp/spawn.test.ts` (new or extend if it exists)

### Implementation steps
1. In `assistant/src/tools/acp/spawn.ts`, replace the hard-coded `claude-agent-acp` check (lines 13-64) with a small registry `ADAPTER_NPM_PACKAGES: Record<string, string>` mapping `"claude-agent-acp" → "@agentclientprotocol/claude-agent-acp"` and `"codex-acp" → "@zed-industries/codex-acp"`.
2. Rewrite `checkAdapterVersion(command)` to:
   - Look up the npm package name in the registry; if not present, return `null` immediately (skip check for unknown adapters).
   - Run both `npm ls -g --json <pkg>` and `npm view <pkg> version` with a 5-second timeout each (use `AbortController` + `setTimeout` against `execFile`).
   - On any error or timeout, return `null` (best-effort).
   - Cache the resolved outcome — including `null` for "skipped" — keyed by command in a process-lifetime map so retries don't reprobe.
3. Change the call site (lines 101-110): if `versionInfo` indicates outdated, append a non-blocking warning string to the success payload's `content` after the JSON, AND keep `isError: false`. Never block the spawn on an outdated check. Update the warning text to use the new `@agentclientprotocol/...` package name.
4. Add tests in `assistant/src/tools/acp/spawn.test.ts`:
   - Synthesize an `execFile` failure → assert spawn proceeds without warning.
   - Synthesize an outdated version → assert spawn proceeds AND warning appears in result content.
   - Synthesize an unknown command → assert no check is performed.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/tools/acp/spawn.test.ts` passes including new cases.
- No spawn that previously succeeded now fails because of the version check.

## PR 3: Add acp_steer tool to redirect running sessions

### Depends on
None

### Branch
acp-codex-claude/pr-3-acp-steer

### Title
feat(acp): add acp_steer tool to interrupt and redirect running sessions

### Files
- `assistant/src/tools/acp/steer.ts` (new)
- `assistant/src/tools/acp/steer.test.ts` (new)
- `assistant/src/config/bundled-skills/acp/tools/acp-steer.ts` (new)
- `assistant/src/config/bundled-skills/acp/TOOLS.json`
- `assistant/src/config/bundled-tool-registry.ts`

### Implementation steps
1. Create `assistant/src/tools/acp/steer.ts` exporting `executeAcpSteer(input, _context): Promise<ToolExecutionResult>`. Match the shape of `tools/acp/abort.ts`:
   - Read `input.acp_session_id` and `input.instruction` as strings.
   - Validate both required, return `{ content, isError: true }` if missing.
   - Call `getAcpSessionManager().steer(acpSessionId, instruction)`.
   - On success, return `{ content: JSON.stringify({ acpSessionId, status: "steered", message: "Interrupted in-flight prompt; new instruction is now running." }), isError: false }`.
   - On error (session not found / not running), return `{ content: ..., isError: true }`.
2. Create `assistant/src/config/bundled-skills/acp/tools/acp-steer.ts` mirroring `acp-spawn.ts` (forward to `executeAcpSteer`).
3. Append a new entry to `assistant/src/config/bundled-skills/acp/TOOLS.json`:
   - `name: "acp_steer"`, `category: "orchestration"`, `risk: "medium"`.
   - `description`: lead with "**Interrupts** the in-flight prompt of a running ACP session and replaces it with `instruction`. Use to redirect the agent (e.g., 'stop, do X instead'). For follow-up work that should happen *after* the current task, do NOT use this — wait for `acp_session_completed` and `acp_spawn` again."
   - `input_schema`: required `acp_session_id` and `instruction`, plus optional `activity` like other entries.
   - `executor: "tools/acp-steer.ts"`, `execution_target: "host"`.
4. Register the executor in `assistant/src/config/bundled-tool-registry.ts` next to the existing `acp_*` entries.
5. Add `assistant/src/tools/acp/steer.test.ts` with cases:
   - Happy path: mock `AcpSessionManager.steer` resolves; assert returned content shape.
   - Missing `instruction`: assert `isError: true`.
   - Missing `acp_session_id`: assert `isError: true`.
   - `manager.steer` throws "session not found": assert error surfaces.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/tools/acp/steer.test.ts` passes.
- `acp_steer` appears in TOOLS.json and in the bundled-tool-registry; description explicitly says "Interrupts".
- E2E manual: with a long-running ACP session, calling `acp_steer` cancels the in-flight prompt and starts the new instruction (verify via SSE).

## PR 4: Add shared resolveAcpAgent helper

### Depends on
PR 1

### Branch
acp-codex-claude/pr-4-resolver

### Title
feat(acp): add shared resolveAcpAgent helper with defaults and binary preflight

### Files
- `assistant/src/acp/resolve-agent.ts` (new)
- `assistant/src/acp/resolve-agent.test.ts` (new)

### Implementation steps
1. Create `assistant/src/acp/resolve-agent.ts`. Export `resolveAcpAgent(id: string): ResolveAcpAgentResult` where the result is a discriminated union:
   ```ts
   type ResolveAcpAgentResult =
     | { ok: true; agent: AcpAgentConfig; source: "config" | "default" }
     | { ok: false; reason: "acp_disabled"; hint: string }
     | { ok: false; reason: "unknown_agent"; hint: string; available: string[] }
     | { ok: false; reason: "binary_not_found"; hint: string; agent: AcpAgentConfig; source: "config" | "default" };
   ```
2. Implementation:
   - `getConfig().acp.enabled` false → `acp_disabled` with hint `"Set 'acp.enabled': true in ~/.vellum/workspace/config.json (or via the runtime config endpoint)."`.
   - `agent = userAgents[id] ?? DEFAULT_ACP_AGENT_PROFILES[id]`. If still undefined → `unknown_agent` with `available` listing the merged keyset.
   - Resolve binary via `Bun.which(agent.command)`. If null → `binary_not_found` with hint from `DEFAULT_AGENT_INSTALL_HINTS[agent.command] ?? `Install '${agent.command}' and ensure it is on PATH.``.
   - Otherwise `ok: true` with `source` set based on whether the user provided the entry.
3. Export a helper `listAcpAgents(): { enabled: boolean; agents: AcpAgentEntry[] }` used by `acp_list_agents` in PR 5. Each entry: `{ id, command, description, source, available, unavailableReason?, setupHint? }`.
4. Add `assistant/src/acp/resolve-agent.test.ts`:
   - acp disabled → `acp_disabled`.
   - id present in user config → `ok: true` with `source: "config"` (user wins over default).
   - id present only in defaults → `ok: true` with `source: "default"`.
   - id missing both → `unknown_agent` and `available` includes both default ids.
   - binary not on PATH → `binary_not_found` with the right install hint (mock `Bun.which`).
   - `listAcpAgents()` shape: includes both defaults when user config is empty; merges correctly when user overrides.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/acp/resolve-agent.test.ts` passes.
- No other module imports `resolve-agent.ts` yet (consumed in PRs 5, 6, 7).

## PR 5: Add acp_list_agents tool with availability hints

### Depends on
PR 3, PR 4

### Branch
acp-codex-claude/pr-5-acp-list-agents

### Title
feat(acp): add acp_list_agents tool with availability hints

### Files
- `assistant/src/tools/acp/list-agents.ts` (new)
- `assistant/src/tools/acp/list-agents.test.ts` (new)
- `assistant/src/config/bundled-skills/acp/tools/acp-list-agents.ts` (new)
- `assistant/src/config/bundled-skills/acp/TOOLS.json`
- `assistant/src/config/bundled-tool-registry.ts`

### Implementation steps
1. Create `assistant/src/tools/acp/list-agents.ts` exporting `executeAcpListAgents(_input, _context): Promise<ToolExecutionResult>`. Body:
   - Call `listAcpAgents()` from `acp/resolve-agent.ts`.
   - If `enabled: false`, return `{ content: JSON.stringify({ enabled: false, hint: "Set 'acp.enabled': true to use ACP agents." }), isError: false }`.
   - Otherwise return `{ content: JSON.stringify({ enabled, agents }), isError: false }`. Include the merge-source label (`config` vs `default`), `available`, `unavailableReason`, `setupHint` for each.
2. Create `assistant/src/config/bundled-skills/acp/tools/acp-list-agents.ts` mirroring the steer wrapper.
3. Append entry to `bundled-skills/acp/TOOLS.json`:
   - `name: "acp_list_agents"`, `category: "orchestration"`, `risk: "low"`.
   - `description`: "Lists ACP coding agents available to spawn. Each entry includes whether ACP is enabled, whether the agent's binary is on PATH, and an install command if not. Use this to decide between 'claude' and 'codex' or to surface setup steps to the user."
   - `input_schema`: optional `activity` only.
   - `executor: "tools/acp-list-agents.ts"`, `execution_target: "host"`.
4. Register the executor in `bundled-tool-registry.ts`.
5. Add `assistant/src/tools/acp/list-agents.test.ts`:
   - Disabled → returns `{ enabled: false, hint }`.
   - Enabled, no user config → both defaults present, marked as `source: "default"`, `available` based on mocked `Bun.which`.
   - Enabled, user overrides claude → claude shows `source: "config"` with the user's command.
   - Unavailable agent surfaces `setupHint` from `DEFAULT_AGENT_INSTALL_HINTS`.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/tools/acp/list-agents.test.ts` passes.
- TOOLS.json contains both `acp_steer` (from PR 3) and `acp_list_agents` entries; no JSON syntax error.

## PR 6: Use shared resolver in acp_spawn tool

### Depends on
PR 2, PR 4

### Branch
acp-codex-claude/pr-6-spawn-tool

### Title
refactor(acp): use shared resolver in acp_spawn tool

### Files
- `assistant/src/tools/acp/spawn.ts`
- `assistant/src/tools/acp/spawn.test.ts`

### Implementation steps
1. In `assistant/src/tools/acp/spawn.ts`, replace the inline `agentConfig` lookup (lines 77-89 in current code, post-PR-2) with a call to `resolveAcpAgent(agent)`.
2. Map the resolver result:
   - `acp_disabled` → `{ content: hint, isError: true }`.
   - `unknown_agent` → `{ content: \`Unknown agent "${agent}". Available: ${available.join(", ")}.\`, isError: true }` (preserve existing message shape).
   - `binary_not_found` → `{ content: \`${agent.command} is not on PATH. ${hint}\`, isError: true }`.
   - `ok: true` → continue with `result.agent` as the existing `agentConfig` variable.
3. Drop the now-unused `config.acp.enabled` and `config.acp.agents[agent]` checks; the resolver covers them.
4. Update tests in `spawn.test.ts`:
   - Add a test for the missing-binary path (mock `Bun.which` to return null in the resolver).
   - Add a test for default-profile fallback when user config is empty.
   - Existing tests for unknown-agent and acp-disabled continue to pass against the new error messages.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/tools/acp/spawn.test.ts` passes.
- E2E manual: with `agents: {}` and `acp.enabled: true`, `acp_spawn { agent: "codex" }` resolves the default profile and either spawns `codex-acp` or returns the install hint if missing.

## PR 7: Use shared resolver in /v1/acp/spawn route

### Depends on
PR 4

### Branch
acp-codex-claude/pr-7-spawn-route

### Title
refactor(acp): use shared resolver in /v1/acp/spawn HTTP route

### Files
- `assistant/src/runtime/routes/acp-routes.ts`
- `assistant/src/runtime/routes/acp-routes.test.ts` (new or extend if exists)

### Implementation steps
1. In `assistant/src/runtime/routes/acp-routes.ts`'s `/v1/acp/spawn` handler (lines 40-95), replace the inline `config.acp.enabled` + `config.acp.agents[body.agent]` checks with `resolveAcpAgent(body.agent)`.
2. Map the resolver result to HTTP errors:
   - `acp_disabled` → `httpError("BAD_REQUEST", hint, 400)`.
   - `unknown_agent` → `httpError("BAD_REQUEST", \`Unknown agent "${body.agent}". Available: ${available.join(", ")}.\`, 400)`.
   - `binary_not_found` → `httpError("FAILED_DEPENDENCY", \`${agent.command} is not on PATH. ${hint}\`, 424)`.
   - `ok: true` → use `result.agent` instead of the previous `agentConfig`.
3. Add or extend route tests to cover all three failure paths. If no test file exists, create `assistant/src/runtime/routes/acp-routes.test.ts` and stub the daemon dependencies as other route tests do.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- Route tests pass.
- The HTTP error responses include the same install-hint text the LLM tool sees, so other gateway/CLI clients of `POST /v1/acp/spawn` get the same DX.

## PR 8: Update bundled SKILL.md for Codex setup, package rename, and steer guidance

### Depends on
PR 3, PR 5

### Branch
acp-codex-claude/pr-8-skill-md

### Title
docs(acp): update SKILL.md for Codex setup, package rename, and steer guidance

### Files
- `assistant/src/config/bundled-skills/acp/SKILL.md`

### Implementation steps
1. Reframe the opening paragraph from "Claude Code adapter" to "external coding agents" (Claude Code, Codex, etc., via ACP).
2. Replace every reference to `@zed-industries/claude-agent-acp` with `@agentclientprotocol/claude-agent-acp` (the old name is deprecated). Update the install command in the "First-time setup" section.
3. Add a new "Codex setup" section after "First-time setup":
   - Install: `npm i -g @zed-industries/codex-acp` (provides the `codex-acp` binary).
   - Codex CLI itself (≥ 0.111) must be on PATH; install via OpenAI's distribution channel of choice.
   - Auth: the `codex-acp` adapter inherits whatever auth the underlying `codex` CLI uses. Typical flows: `codex login` (OAuth), or `CODEX_API_KEY`, or `OPENAI_API_KEY` env var. Mention that the assistant will surface the install hint via `acp_list_agents` if `codex-acp` isn't on PATH.
4. Update the "Critical: correct agent command" section: `claude-agent-acp` and `codex-acp` are the two supported adapter binaries today. Default profiles ship out-of-box; users only need `agents.<id>` if they want to override.
5. Add a new "When to use acp_steer vs acp_spawn" section:
   - `acp_steer` interrupts the in-flight prompt — use it to course-correct ("stop, do X instead").
   - For follow-ups after the current task ("also do Y when you're done"), wait for `acp_session_completed` and call `acp_spawn` again. Queued follow-ups in the same session are not yet supported.
6. Add a short "Discoverability" section pointing at `acp_list_agents` for "what's set up, what's missing".
7. Add a "Working directory" tip: default to the conversation's cwd; for risky/parallel work, create a git worktree first via shell tool and pass that path as `cwd` to `acp_spawn`.

### Acceptance criteria
- SKILL.md no longer references `@zed-industries/claude-agent-acp` (deprecated).
- Codex setup including auth is documented.
- `acp_steer` is unambiguously labeled interruptive in the doc.
- `bunx tsc --noEmit` passes (no code change here, but the bundled-skill emit script may regenerate manifests).

## PR 9: Release notes migration for ACP Codex + steer

### Depends on
PR 5, PR 8

### Branch
acp-codex-claude/pr-9-release-notes

### Title
feat(workspace): release notes migration for ACP Codex + steer

### Files
- `assistant/src/workspace/migrations/053-release-notes-acp-codex.ts` (new)
- `assistant/src/workspace/migrations/registry.ts`
- `assistant/src/__tests__/workspace-migration-053-release-notes-acp-codex.test.ts` (new)

### Implementation steps
1. Create `assistant/src/workspace/migrations/053-release-notes-acp-codex.ts`. Use prior release-note migrations as templates (e.g. `043-release-notes-latex-rendering.ts`, `045-release-notes-meet-avatar.ts`, `049-release-notes-default-sonnet.ts`). The migration must:
   - Export a `WorkspaceMigration` object with `id: "053-release-notes-acp-codex"`.
   - In `run()`, locate `<workspace>/UPDATES.md`, read existing content, short-circuit if `<!-- release-note-id:053-release-notes-acp-codex -->` is already present (idempotency rule from CLAUDE.md § Release Update Hygiene).
   - Otherwise append a release note block with the marker. Body content:
     - Default profiles for `claude` and `codex` ship with the assistant; available **after enabling ACP and installing the corresponding adapter**.
     - `acp_steer` lets the assistant interrupt and redirect a running session.
     - Setup: set `acp.enabled: true`; `npm i -g @zed-industries/codex-acp` and/or `npm i -g @agentclientprotocol/claude-agent-acp`. The assistant will surface install hints if a binary is missing.
     - Note: live step-by-step progress is not yet rendered in the macOS app; the agent's final response lands in chat when it completes.
2. Append the new migration to `WORKSPACE_MIGRATIONS` in `assistant/src/workspace/migrations/registry.ts` (do NOT reorder existing entries).
3. Add `assistant/src/__tests__/workspace-migration-053-release-notes-acp-codex.test.ts` mirroring existing release-note migration tests:
   - Migration appends the expected text to UPDATES.md when run on a fresh workspace.
   - Re-running with the marker already present is a no-op (idempotent).
   - Re-running without the marker (e.g., user deleted UPDATES.md) re-appends.

### Acceptance criteria
- `bunx tsc --noEmit` passes.
- `bun test src/__tests__/workspace-migration-053-release-notes-acp-codex.test.ts` passes.
- `WORKSPACE_MIGRATIONS` array length grew by exactly 1 with the new migration appended at the end.

</details>

**Comment by siddseethepalli** [comment]
Addressed in https://github.com/vellum-ai/vellum-assistant/pull/29970

File: assistant/src/config/acp-defaults.ts
[the remark this answers is not in the document]
Comment (siddseethepalli) [reply assistant/src/config/acp-defaults.ts:1]: Addressed in https://github.com/vellum-ai/vellum-assistant/pull/29970

File: assistant/src/config/acp-defaults.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  claude: Object.freeze({
+    command: "claude-agent-acp",
+    args: [],
+    description: "Claude Code (via @agentclientprotocol/claude-agent-acp)",
+  }),
+  codex: Object.freeze({
+    command: "codex-acp",
+    args: [],
+    description: "OpenAI Codex CLI (via @zed-industries/codex-acp)",
+  }),
```
[the remark this answers is not in the document]
Comment (siddseethepalli) [reply assistant/src/config/acp-defaults.ts:25]: Addressed in https://github.com/vellum-ai/vellum-assistant/pull/29970

File: assistant/src/config/acp-defaults.test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      DEFAULT_AGENT_INSTALL_HINTS["new-binary"] = "npm i -g foo";
+    };
+    const _assignNewProfile: () => void = () => {
+      // @ts-expect-error — DEFAULT_ACP_AGENT_PROFILES has a Readonly index signature
+      DEFAULT_ACP_AGENT_PROFILES.newAgent = { command: "x", args: [] };
+    };
+    // The assertions live in the @ts-expect-error comments above; this test
+    // exists to surface a type-check failure if the readonly contract regresses.
+    expect(_assignNewKey).toBeFunction();
+    expect(_assignNewProfile).toBeFunction();
```
[the remark this answers is not in the document]
Comment (siddseethepalli) [reply assistant/src/config/acp-defaults.test.ts:82]: Addressed in https://github.com/vellum-ai/vellum-assistant/pull/29970
